### PR TITLE
Fix use explicit type and nullable-annotated array syntax generation

### LINKED ISSUE #37877: IDE0008 doesn't account for nullable
**Version Used**: 
3.3.0-beta2-19401-05+8ed56b826700926009731ca4d3de0a4e4e652969

**Steps to Reproduce**:
Copy/paste this into new console app:
```C#
#nullable enable
class Program
{
    private static int[]? s_data;

    static void Main()
    {
        var data = s_data;
    }
}
```
then run the IDE0008 refactoring on the `var`.

**Expected Behavior**:
Produces:
```C#
#nullable enable
class Program
{
    private static int[]? s_data;

    static void Main()
    {
        int[]? data = s_data;
    }
}
```

**Actual Behavior**:
Produces:
```C#
#nullable enable
class Program
{
    private static int[]? s_data;

    static void Main()
    {
        int[] data = s_data;
    }
}
```
resulting in the warning:
```
warning CS8600: Converting null literal or possible null value to non-nullable type.
```

### PR DESCRIPTION
**Pull request by jasonmalinowski**
Fixes #37877.

### CODE REVIEW

File: src/Workspaces/CSharp/Portable/Extensions/ITypeSymbolExtensions.TypeSyntaxGeneratorVisitor.cs [hunk 1]
```diff
@@ -62,25 +62,51 @@ public override TypeSyntax VisitArrayType(IArrayTypeSymbol symbol)
...
+                ITypeSymbol underlyingType = symbol;
+
+                while (underlyingType is IArrayTypeSymbol innerArray)
                 {
-                    underlyingNonArrayType = ((IArrayTypeSymbol)underlyingNonArrayType).ElementType;
+                    underlyingType = innerArray.GetElementTypeWithAnnotatedNullability();
+
+                    if (underlyingType.GetNullability() == NullableAnnotation.Annotated)
+                    {
+                        // If the inner array we just moved to is also nullable, then
```
Comment (jasonmalinowski) [inline src/Workspaces/CSharp/Portable/Extensions/ITypeSymbolExtensions.TypeSyntaxGeneratorVisitor.cs]: @dotnet/roslyn-compiler: where is the spec for how we are (or are not) reordering array ranks when there are nullable annotations? I played around with a bit and arrived at this but an actual spec would be wonderful.

File: src/EditorFeatures/CSharpTest/CodeRefactorings/UseExplicitOrImplicitType/UseExplicitTypeRefactoringTests.cs [hunk 2]
```diff
@@ -293,6 +293,134 @@ void M()
...
+
+        [Fact]
+        public async Task TestWithTopLevelAndNestedArrayNullability1()
+        {
+            var code = @"
+#nullable enable
+
+class C
+{
+    private static string?[]?[,]? s_data;
```
Comment (genlu) [inline src/EditorFeatures/CSharpTest/CodeRefactorings/UseExplicitOrImplicitType/UseExplicitTypeRefactoringTests.cs]: O_O

File: src/Workspaces/CSharp/Portable/Extensions/ITypeSymbolExtensions.TypeSyntaxGeneratorVisitor.cs (hunk 1 above)
Comment (genlu) [reply src/Workspaces/CSharp/Portable/Extensions/ITypeSymbolExtensions.TypeSyntaxGeneratorVisitor.cs]: This is quite confusing indeed.

---
In reply to: [312684703](https://github.com/dotnet/roslyn/pull/37892#discussion_r312684703) [](ancestors = 312684703)

Comment (sharwell) [reply src/Workspaces/CSharp/Portable/Extensions/ITypeSymbolExtensions.TypeSyntaxGeneratorVisitor.cs]: @jasonmalinowski We need to make sure our code fix for CS8600 accounts for this

Comment (sharwell) [inline src/Workspaces/CSharp/Portable/Extensions/ITypeSymbolExtensions.TypeSyntaxGeneratorVisitor.cs]:
```suggestion
                        // 🐉 If the inner array we just moved to is also nullable, then
```

Comment (genlu) [reply src/Workspaces/CSharp/Portable/Extensions/ITypeSymbolExtensions.TypeSyntaxGeneratorVisitor.cs]: What does the dragon mean?

Comment (jasonmalinowski) [reply src/Workspaces/CSharp/Portable/Extensions/ITypeSymbolExtensions.TypeSyntaxGeneratorVisitor.cs]: Won't fix.

Comment (jasonmalinowski) [reply src/Workspaces/CSharp/Portable/Extensions/ITypeSymbolExtensions.TypeSyntaxGeneratorVisitor.cs]: (I storngly disagree with use of emoji in general as demonstrated by @genlu: those can be confused as the meaning. Text less so.)

Comment (sharwell) [reply src/Workspaces/CSharp/Portable/Extensions/ITypeSymbolExtensions.TypeSyntaxGeneratorVisitor.cs]: It's a fairly common term in programming sourced originally from https://en.wikipedia.org/wiki/Here_be_dragons

File: src/Workspaces/CSharp/Portable/Extensions/ITypeSymbolExtensions.TypeSyntaxGeneratorVisitor.cs [hunk 3]
```diff
@@ -62,25 +62,51 @@ public override TypeSyntax VisitArrayType(IArrayTypeSymbol symbol)
...
+                    }
                 }
 
-                var elementTypeSyntax = underlyingNonArrayType.GenerateTypeSyntax();
+                var elementTypeSyntax = underlyingType.GenerateTypeSyntax();
                 var ranks = new List<ArrayRankSpecifierSyntax>();
 
                 var arrayType = symbol;
-                while (arrayType != null)
+                while (arrayType != null && !arrayType.Equals(underlyingType))
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/Extensions/ITypeSymbolExtensions.TypeSyntaxGeneratorVisitor.cs]: any concern about the right nullability comparison happening here?

Comment (jasonmalinowski) [reply src/Workspaces/CSharp/Portable/Extensions/ITypeSymbolExtensions.TypeSyntaxGeneratorVisitor.cs]: It'll be doing the non-nullable-aware compare but the "interesting" part of the comparison anyways is the actual array depth. Switching to nullable aware would be the same behavior.

Comment (CyrusNajmabadi) [reply src/Workspaces/CSharp/Portable/Extensions/ITypeSymbolExtensions.TypeSyntaxGeneratorVisitor.cs]: SGTM!

**Pull request merged**